scripts/Stretch.py

```diff
@@ -0,0 +1,357 @@
+            vPosAfter = np.resize(np.roll(vPos,-i-1,0),(iextra,2))
+            good_triangle = True
+            dp = ((vPos[i]-vPosAfter)**2).sum(1)
+            if np.amax(dp) < d1*d1:
+                good_triangle = False
+            else:
+                i3 = np.argmax(dp>=d1*d1)
+            if good_triangle:
+                vPosBefore = np.resize(np.roll(vPos,-i,0)[::-1],(iextra,2))
+                dp = ((vPos[i]-vPosBefore)**2).sum(1)
+                if np.amax(dp) < d1*d1:
+                    good_triangle = False
+                else:
+                    i1 = np.argmax(dp>=d1*d1)
+            if good_triangle:
+                rd=((vPosAfter[i3]-vPosBefore[i1])**2).sum(0)
+                r=((vPos[i]-vPosBefore[i1])*(vPosAfter[i3]-vPosBefore[i1])).sum(0)
+                if np.fabs(r) < 1000.0*np.fabs(rd):
+                    r /= rd
+                else:
+                    r = 0.5
+                pp=vPosBefore[i1] + r*(vPosAfter[i3]-vPosBefore[i1])
+                dpp=np.sqrt(((pp-vPos[i])**2).sum(0))
+                if dpp > 0.001:
+                    vTrans[i] = vPos[i] - (d4/dpp)*(pp-vPos[i])
+
+            i = i + 1
+        return
+
+    def wideTurn(self,vPos,vTrans):
+        '''
+        We have to select three points in order to form a triangle
+        These three points should be far enough from each other to have
+        a reliable estimation of the orientation of the current turn
+        '''
+        d1 = self.line_width / 2.0
+        d4 = self.stretch
+        i1 = 0
+        i = 1
+        i3 = 2
+        while i+1<len(vPos):
+            good_triangle = True
+            dp = ((vPos[i]-vPos[i+1:])**2).sum(1)
+            if np.amax(dp) < d1*d1:
+                good_triangle = False
+            else:
+                i3 = i+1 + np.argmax(dp>=d1*d1)
+            if good_triangle:
+                dp = ((vPos[i]-vPos[i-1::-1])**2).sum(1)
+                if np.amax(dp) < d1*d1:
+                    good_triangle = False
+                else:
+                    i1 = i-1 - np.argmax(dp>=d1*d1)
+            if good_triangle:
+                rd=((vPos[i3]-vPos[i1])**2).sum(0)
+                r=((vPos[i]-vPos[i1])*(vPos[i3]-vPos[i1])).sum(0)
+                if np.fabs(r) < 1000.0*np.fabs(rd):
+                    r /= rd
+                else:
+                    r = 0.5
+                pp=vPos[i1] + r*(vPos[i3]-vPos[i1])
+                dpp=np.sqrt(((pp-vPos[i])**2).sum(0))
+                if dpp > 0.001:
+                    vTrans[i] = vPos[i] - (d4/dpp)*(pp-vPos[i])
+            i = i + 1
+        return
+
+    def pushWall(self,vPos,vTrans):
+        d2 = self.line_width / 2.0
+        d4 = self.stretch
+        mrot = np.array([[0,-1],[1,0]]) # Quarter turn
+        for i in range(len(vPos)):
+            i1 = i
+            i2 = i+1
+            if i2 == len(vPos):
+                i2 = i-1
+            xm = vPos[i1]
+            xperp = np.dot(mrot,vPos[i2]-vPos[i1])
```

Comment:
So this rotates the direction vector after `i` 90 degrees counter-clockwise. The resulting direction vector is scaled and then added to the vector at `i`. The problem is then that the direction vector before `i` is not taken into account. If the angle is more than 90 degrees, then this would result in the vertex being moved towards the new line, making a smaller turn rather than a larger one. To properly produce the normal vector of a vertex, you need to take both lines into account. It's commonly done by taking the average of the two direction vectors going away from the vertex (and guarding against zero-vectors then). That's also much simpler than computing a perpendicular.